Add modules to compute hypervolume.

HideakiImamura opened the pull request:
## Motivation
For multiobjective optimization, we need to compute hypervolume for the given objective set in some cases. This PR adds some classes to compute hypervolume.

## Description of the changes
- Add a base class to compute hypervolume.
- Add a class to compute hypervolume for 2-dimensional cases.
- Add a class to compute hypervolume for any dimensional cases based on the WFG algorithm.
- Add some utility functions.
- Add unit tests.

## Note
- We need some discussions for user-interface to compute hypervolume.
- We need to speed up the added WFG class. Specifically, we can implement the "slicing" technique explained in the [original paper](https://ieeexplore.ieee.org/document/5766730). This can be done in another PR.

## TODO
- [x] Benchmark accuracy comparison experiment with `pagmo2`.
- [x] Benchmark speed comparison experiment with `pagmo2`

HideakiImamura commented:
## Benchmark accuracy comparison experiment with `pagmo`.
I compared the hypervolume computing module introduced in this PR with that of `pagmo`. The hypervolume for each test case calculated by both modules is allowed up to a relative error of 1%.

### How to generate test cases.
I used the Dockerfile in [this `pagmo_on_docker` repository](https://github.com/HideakiImamura/pagmo_on_docker). Adjusted some parameters in `pagmo_on_docker/maker.cpp` as follows.
- `n`: the number of test cases.
- `d`: the dimension of test cases.
-  Set `(n, d) = (1000, 2), (100, 3), (30, 10)`.
I also changed the generated file of test cases as `test_case_2d.csv` and so on.

According to the README.md of the `pagmo_on_docker`, I built the docker image, run container, compiled `maker.cpp`. and executed the compiled file.  Then I got some test files. Each row of the test files is `solution_set, reference_point, hypervolume_computed_by_pagmo`.

### How to test.
By using the generated test cases, I confirmed the introduced module can compute hypervolume with sufficient accuracy for `pagmo` based on the following test code.
```python
from ast import literal_eval
import csv

import numpy as np

import optuna


def test_file_wfg_2d() -> None:
    with open('./test_cases_2d.csv') as f:
        reader = csv.reader(f, delimiter=' ')
        for row in reader:
            s = np.asarray(literal_eval(row[0]))
            r = np.asarray(literal_eval(row[1]))
            v = float(row[2])
            assert np.isclose(v, optuna.multi_objective.hypervolume.WFG().compute(s, r), rtol=1e-2)


def test_file_wfg_3d() -> None:
    with open("./test_cases_wfg_3d.csv") as f:
        reader = csv.reader(f, delimiter=" ")
        for i, row in enumerate(reader):
            s = np.asarray(literal_eval(row[0]))
            r = np.asarray(literal_eval(row[1]))
            v = float(row[2])
            assert np.isclose(v, optuna.multi_objective.hypervolume.WFG().compute(s, r), rtol=1e-2)


def test_file_wfg_10d() -> None:
    with open(./test_cases_wfg.csv") as f:
        reader = csv.reader(f, delimiter=" ")
        for i, row in enumerate(reader):
            s = np.asarray(literal_eval(row[0]))
            r = np.asarray(literal_eval(row[1]))
            v = float(row[2])
            assert np.isclose(v, optuna.multi_objective.hypervolume.WFG().compute(s, r), rtol=1e-2)

if __name__ == '__main__':
    test_file_wfg_2d()
    test_file_wfg_3d()
    test_file_wfg_10d()
```
Executing this code does not raise any errors, so I confirmed the computing module of this PR can compute hypervolume with sufficient accuracy.

hvy commented on optuna/multi_objective/hypervolume/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    INCOMPARABLE = 3
+
+
+def dom(point1: np.ndarray, point2: np.ndarray, dim_bound: Optional[int] = None) -> DomRelation:
```
How about a clearer name e.g. similar to https://github.com/optuna/optuna/blob/master/optuna/multi_objective/trial.py#L337 ?

hvy commented on optuna/multi_objective/hypervolume/__init__.py:
```diff
@@ -0,0 +1,98 @@
+
+from optuna.multi_objective.hypervolume.base import BaseHypervolume  # NOQA
+from optuna.multi_objective.hypervolume.exact_2d import Exact2d  # NOQA
+from optuna.multi_objective.hypervolume.wfg import WFG  # NOQA
```
Are these imports located at the bottom for some specific reason? Could be worth adding a comment.

HideakiImamura replied:
I will split the implementation of `compute_wpoints_volume` and `domination functions` with `__init__.py`.

hvy commented on optuna/multi_objective/hypervolume/base.py:
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+
+
+class BaseHypervolume(object, metaclass=abc.ABCMeta):
```
Do you think it's better to make the classes/the hypervolume directory public? If it's only intended for MOTPE for the time being, it could be private and I also think we can reduce the number of "validations".

HideakiImamura replied:
I agree with the idea of making the hypervolume class private. This would reduce the validation function and the overall amount of code.

hvy commented on optuna/multi_objective/hypervolume/wfg.py:
```diff
@@ -0,0 +1,83 @@
+    """
+
+    def __init__(self) -> None:
+        self._r = np.ndarray(shape=())
```
How come the reference point is kept as a member (while the solution set isn't)? Might be cleaner just to make it stateless?

HideakiImamura replied:
Yes, because a solution set can change with each internal private function call, but the reference point is immutable once given to the user.

hvy commented on optuna/multi_objective/hypervolume/exact_2d.py:
```diff
@@ -0,0 +1,35 @@
+        if solution_set.shape[0] == 1:
+            return compute_2points_volume(solution_set[0], reference_point)
+
+        sorted(solution_set, key=lambda p: p[1])
```
The return value is not used here, does it still work?

hvy replied:
Maybe shuffle the solution set in the tests?

hvy replied:
By the way, you don't have to convert it to a list.
```suggestion
        solution_set = solution_set[solution_set[:, 1].argsort()]
```

hvy commented on optuna/multi_objective/hypervolume/exact_2d.py:
```diff
@@ -0,0 +1,35 @@
+    def __init__(self) -> None:
+        pass
+
+    def compute(self, solution_set: np.ndarray, reference_point: np.ndarray) -> float:
```
Just for the record, tried with creating a tiled array + masking but it was a lot slower than your looped version. Likely because of the n^2 memory allocation. For those who are interested.
```python
        solution_set = solution_set[solution_set[:, 0].argsort()]

        n = solution_set.shape[0]
        widths = reference_point[1] - solution_set[:, 1]
        widths_masked = np.repeat(widths[:, None], n, axis=1)
        widths_masked = widths_masked.astype(np.float32)
        widths_masked[np.tril_indices(n, k=-1)] = -np.inf
        widths = widths_masked.max(axis=0)

        heights =  np.append(solution_set[1:, 0], reference_point[0]) - solution_set[:, 0]

        return np.sum(widths * heights)
```

HideakiImamura replied:
Thanks for the investigation. The logic of the function is left as it is, but I've modified the naming conventions for the variables based on you.

HideakiImamura commented on optuna/multi_objective/hypervolume/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    """
+
+    P1_DOM_P2 = 0
+    P2_DOM_P1 = 1
```
Thanks for the comment. The logic can be improved by replacing `dom` and `DomRelation` by a single function. I will do that.

HideakiImamura commented on optuna/multi_objective/hypervolume/wfg.py:
```diff
@@ -0,0 +1,83 @@
+                if p_is_not_inserted:
+                    assert all(p == np.maximum(s[i], p))
+                    limited_s.append(np.maximum(s[i], p))
+                    p_is_not_inserted = False
```
I have noticed that the logic is wrong. I fixed them.

HideakiImamura commented on optuna/multi_objective/hypervolume/base.py:
```diff
@@ -0,0 +1,47 @@
+        raise NotImplementedError
+
+    @staticmethod
+    def _validate(solution_set: np.ndarray, reference_point: np.ndarray) -> None:
```
Yes, it is a same assumption with `pagmo`.

HideakiImamura commented:
@hvy Thanks for your reviews.

> Also, just wondering but is the user of these classes expected to have "normalized" the points such that it's always a minimization problem?

Yes, I think the interface of this module should be simpler, so the inputs of these classes to be "normalized". I think the high-level class which uses the hypervolume computation modules should be responsible for converting a maximization problem to a minimization problem. What do you think?

HideakiImamura commented:
PTAL.

hvy commented on optuna/multi_objective/_hypervolume/exact_2d.py:
```diff
@@ -0,0 +1,33 @@
+        )
+        heights = np.hstack([solution_set[1:, 1], reference_point[1]]) - solution_set[:, 1]
+        v = float(np.sum(widths * heights))
+        return v
```
I think 
```python
widths = reference_point[0] - solution_set[:, 0]
widths = np.asarray([widths[: i + 1].max() for i in range(widths.size)])
edges = np.append(solution_set[1:, 1], reference_point[1]) - solution_set[:, 1]
return float(np.sum(widths * edges))
```

HideakiImamura replied:
Thanks for the suggestion! I check the speed using the above `test_case_2d` benchmark. Your suggestion improves speed!
| 5daeec8 | With your suggestion  | 
|---|---|
|  8.90 sec | 7.39 sec  |

HideakiImamura commented:
## Benchmark speed comparison experiment with `pagmo2`.
I compared the hypervolume computing module introduced in this PR with that of `pagmo2`. We compare the speed to compute the hypervolume calculated by both modules.

### How to generate test cases.
Same as https://github.com/optuna/optuna/pull/1537#issuecomment-661634529.

### How to test.
By using the generated test cases, I confirmed the introduced module can compute hypervolume with sufficient speed for `pagmo2` based on the following test code.
```python
from ast import literal_eval
import csv
import time

import numpy as np

import optuna


def test_file_wfg_2d() -> None:
    with open('./test_cases_wfg_2d.csv') as f:
        with open('./comp_time_2d.csv', 'w') as g:
            reader = csv.reader(f, delimiter=' ')
            for i, row in enumerate(reader):
                if (i + 1) // 100 == 0:
                    if (i + 1) % 10 != 0:
                        continue
                elif (i + 1) % 100 != 0:
                    continue
                s = np.asarray(literal_eval(row[0]))
                r = np.asarray(literal_eval(row[1]))
                t = float(row[3]) / 1e+6
                start = time.time()
                optuna.multi_objective._hypervolume.WFG().compute(s, r)
                elapsed = time.time() - start
                g.write("{} {} {}\n".format(i + 1, t, elapsed))


def test_file_wfg_3d() -> None:
    with open("./test_cases_wfg_3d.csv") as f:
        with open('./comp_time_3d.csv', 'w') as g:
            reader = csv.reader(f, delimiter=" ")
            for i, row in enumerate(reader):
                if (i + 1) // 100 == 0:
                    if (i + 1) % 10 != 0:
                        continue
                elif (i + 1) % 100 != 0:
                    continue
                s = np.asarray(literal_eval(row[0]))
                r = np.asarray(literal_eval(row[1]))
                t = float(row[3]) / 1e+6
                start = time.time()
                optuna.multi_objective._hypervolume.WFG().compute(s, r)
                elapsed = time.time() - start
                g.write("{} {} {}\n".format(i + 1, t, elapsed))


def test_file_wfg_10d() -> None:
    with open("./test_cases_wfg_10d.csv") as f:
        with open('./comp_time_10d.csv', 'w') as g:
            reader = csv.reader(f, delimiter=" ")
            for i, row in enumerate(reader):
                if (i + 1) // 100 == 0:
                    if (i + 1) % 10 != 0:
                        continue
                elif (i + 1) % 100 != 0:
                    continue
                s = np.asarray(literal_eval(row[0]))
                r = np.asarray(literal_eval(row[1]))
                t = float(row[3]) / 1e+6
                start = time.time()
                optuna.multi_objective._hypervolume.WFG().compute(s, r)
                elapsed = time.time() - start
                g.write("{} {} {}\n".format(i + 1, t, elapsed))

if __name__ ==  '__main__'
    test_file_wfg_2d()
    test_file_wfg_3d()
    test_file_wfg_10d()
```
Executing this code generates `comp_time_2d.csv` and so on.

### Results
#### 2-d case
| # of points | pagmo2 [sec] | this PR [sec] |
|---|---|---|
| 10 | 9e-09 | 0.0003311634063720703
| 20 | 8.400000000000001e-08 | 0.00016880035400390625
| 30 | 8.599999999999999e-08 | 0.0002071857452392578
| 40 | 3.9e-08 | 0.0002620220184326172
| 50 | 1.05e-07 | 0.0003292560577392578
| 60 | 3.15e-07 | 0.00036716461181640625
| 70 | 1.49e-07 | 0.00041675567626953125
| 80 | 1.54e-07 | 0.0004601478576660156
| 90 | 1.22e-07 | 0.0005221366882324219
| 100 | 2.03e-07 | 0.0006530284881591797
| 200 | 6.4e-07 | 0.0011811256408691406
| 300 | 2.476e-06 | 0.00197601318359375
| 400 | 2.431e-06 | 0.0023038387298583984
| 500 | 3.505e-06 | 0.002766847610473633
| 600 | 4.9539999999999995e-06 | 0.0032558441162109375
| 700 | 7.183e-06 | 0.004307985305786133
| 800 | 9.034e-06 | 0.004540920257568359
| 900 | 1.1754e-05 | 0.005294084548950195
| 1000 | 1.3089000000000001e-05 | 0.005656719207763672
#### 3-d case
| # of points | pagmo2 [sec] | this PR [sec] |
|---|---|---|
| 10 | 8e-09 | 0.0018620491027832031
| 20 | 3.9e-08 | 0.006571054458618164
| 30 | 2.3e-08 | 0.008652925491333008
| 40 | 4.4e-08 | 0.02123093605041504
| 50 | 3.68e-07 | 0.09856390953063965
| 60 | 2.47e-07 | 0.03000926971435547
| 70 | 1.84e-07 | 0.21518516540527344
| 80 | 1.88e-07 | 0.06515693664550781
| 90 | 1.98e-07 | 0.16269898414611816
| 100 | 3.4899999999999996e-07 | 0.6544809341430664
| 200 | 8.97e-07 | 1.1911752223968506
| 300 | 1.6990000000000002e-06 | 2.1438350677490234
| 400 | 2.754e-06 | 4.408586025238037
| 500 | 4.872e-06 | 5.220775842666626
| 600 | 6.193e-06 | 18.110975980758667
| 700 | 9.037e-06 | 12.598584175109863
| 800 | 1.0419e-05 | 13.725198745727539
| 900 | 1.3006e-05 | 25.77805185317993
| 1000 | 1.7399e-05 | 12.696228981018066
#### 10-d case
| # of points | pagmo2 [sec] | this PR [sec] |
|---|---|---|
| 10 | 3.7999999999999996e-08 | 0.012073993682861328
| 20 | 8.5e-07 | 0.30254220962524414
| 30 | 1.3029999999999998e-06 | 21.63835906982422
| 40 | 2.81e-06 | 112.07284092903137
| 50 | 1.4691e-05 | 3245.8354711532593
| 60 | 1.2356e-05 | 13039.908929109573

toshihikoyanase commented on optuna/multi_objective/_hypervolume/utils.py:
```diff
@@ -0,0 +1,55 @@
+
+    assert isinstance(dim_bound, int)
+
+    for i in range(dim_bound):
+        if point1[i] < point2[i]:
+            return False  # Incompatible or point2 dominates point1
+        elif point1[i] > point2[i]:
+            if any([point1[j] < point2[j] for j in range(i + 1, dim_bound)]):
+                return False  # Incompatible
+            else:
+                return True  # point1 dominates point2
+    return True
```
This logic works fine, but I guess we can simplify this logic as follows:

```suggestion
    if any(point1[i] < point2[i] for i in range(dim_bound)):
        return False
    return True
```

This function returns `False` only if the element of `point1` is smaller than the element of `point2`, so the above code simply checks it first. But, I think we can work on this suggestion in a new PR since the current code is examined by the benchmark tests.

toshihikoyanase commented on optuna/multi_objective/_hypervolume/wfg.py:
```diff
@@ -0,0 +1,82 @@
+class WFG(BaseHypervolume):
+    """Hypervolume calculator for any dimension.
+
+        This class exactly calculates the hypervolume for any dimension by using the WFG algorithm.
+        For detail, see `While, Lyndon, Lucas Bradstreet, and Luigi Barone. "A fast way of
+        calculating exact hypervolumes." Evolutionary Computation, IEEE Transactions on 16.1 (2012)
+        : 86-95.`.
```
I think these lines are indented too deeply compared to the other classes.
```suggestion
    This class exactly calculates the hypervolume for any dimension by using the WFG algorithm.
    For detail, see `While, Lyndon, Lucas Bradstreet, and Luigi Barone. "A fast way of
    calculating exact hypervolumes." Evolutionary Computation, IEEE Transactions on 16.1 (2012)
    : 86-95.`.
```

hvy commented on optuna/multi_objective/_hypervolume/utils.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def _compute_2points_volume(
+    point1: np.ndarray, point2: np.ndarray, dim_bound: Optional[int] = None
```
A comment not particularly for this function but it's probably cleaner to remove `dim_bound` until we actually need it.

HideakiImamura commented:
Thanks for the reviews! PTAL.

toshihikoyanase commented on optuna/multi_objective/_hypervolume/wfg.py:
```diff
@@ -0,0 +1,86 @@
+            left = 0
+            right = 1
+            while left < right < n_points_of_s:
+                if not _dominates_or_equal(
+                    limited_solution_set[right], limited_solution_set[left]
+                ):
+                    left = right
+                    returned_limited_solution_set.append(limited_solution_set[left])
```
These lines are a bit difficult for me to understand, and I want a comment here too.
Intuitively, if `right` is not dominated by `left`, `right` is a part of the Pareto front.

And also, I'm not sure if this logic always extracts the Pareto optimal solutions.
What happens if the `right` dominates `left`?

```suggestion
                if not _dominates_or_equal(
                    limited_solution_set[right], limited_solution_set[left]
                ):
                    # `right` is not dominated by `left`, and it corresponds to a Pareto optimal point.
                    returned_limited_solution_set.append(limited_solution_set[right])
                    left = right
```

toshihikoyanase replied:
For me, the negation of `_dominates_or_equal` is a bit confusing. I prefer the following style, but this is not a request for the change.
```python
                if (limited_solution_set[left] > limited_solution_set[right]).any():
                    # If the `right` is not dominated by `left`
```

HideakiImamura replied:
Thanks for the comment.
>And also, I'm not sure if this logic always extracts the Pareto optimal solutions.
What happens if the right dominates left?

Yes, if the `right` dominates `left`, we have to consider a lot of cases to decide Pareto optimal points. However, the input `solution_set` of `_limit()` method is assumed to be sorted in `_compute_rec()` method (Line 38). Therefore, we don't have to consider such cases.

I left a comment about those points.

toshihikoyanase replied:
Maybe I'm wong, but I think `right` can dominate `left` since the `solution_set` is only sorted by 0-th dimension. It means the order of 1st and later dimensions are undefined. For example, if we have two 2-dimensional vectors like `(0, 1)` and `(0, 2)`, the `right` can be `(0, 1)`, which dominates `left=(0, 2)`. In that case, `left` seems to be included even if it is not a Pareto optimal point. The following is an example input:

```python
>>> _limit(np.array([0, 0, 0]), np.array([[0, 1, 2], [0, 2, 2]]))
array([[0, 1, 2]])
>>> _limit(np.array([0, 0, 0]), np.array([[0, 2, 2], [0, 1, 2]]))
array([[0, 2, 2],
       [0, 1, 2]])
```


This phenomenon actually happens if we slightly modified a test case.

<details>
<summary>For example, if we add one more `(0, 1)` to `test_wfg_2d`:</summary>

```python
def test_wfg_2d() -> None:
    for n in range(2, 30):
        r = n * np.ones(2)
        s = np.asarray([[n - 1 - i, i] for i in range(n)])
        for i in range(n + 1):
            s = np.vstack((s, np.asarray([i, n - i])))
        s = np.vstack(np.asarray((s, [0, 1])))  # This line is added.
        np.random.shuffle(s)
        v = optuna.multi_objective._hypervolume.WFG().compute(s, r)
        assert v == n * n - n * (n - 1) // 2
```

Output 1:
```console
Sorted solution set [[0 1]
 [0 1]
 [0 2]
 [1 0]
 [1 1]
 [2 0]]
Point in _limit [0 1]
Solution_set in _limit [[0 1]
 [0 2]
 [1 1]
 [1 1]
 [2 1]]
Limited solution set [array([0, 1])]  # expected
```

Output 2:
```console
Sorted solution set [[0 1]
 [0 2]
 [0 1]
 [1 0]
 [1 1]
 [2 0]]
Point in _limit [0 1]
Solution set in _limit  [[0 2]
 [0 1]
 [1 1]
 [1 1]
 [2 1]]
Limited solution set [array([0, 2]), array([0, 1])]
```
</details>

To fix this case, we'll need to consider all dimensions when we sort the `solution_set` (c.f., we can use `np.lexsort(np.rot90(solution_set))` as mentioned [here](https://stackoverflow.com/questions/38277143/sort-2d-numpy-array-lexicographically)). But, I think it's OK to accept such edge cases since we don't require exact Pareto front here.

In addition to that, I noticed that the current WFG algorithm may not consider the duplication of points. I think it can be seen if the objective values are integer. What do you think of it?

HideakiImamura replied:
I agree with you. Because the performance degradation will be small due to the miss of "complete" Pareto optimal set in the `_limit()` method, I will fix only the comment in the source code. Thank you for your insightful reviews.

HideakiImamura commented on optuna/multi_objective/_hypervolume/utils.py:
```diff
@@ -0,0 +1,32 @@
+        A boolean value representing whether the point1 dominates or equal to point2.
+    """
+
+    if (point1 < point2).any():
+        return False
+    return True
```
Thanks for the comment. I've noticed the current code is broken. Since we want to implement "point1 dominates point2", the code is fixed as follows. The inequality is flipped against the previous version.
```python
def _dominates_or_equal(point1, point2):
    if (point1 > point2).any():
        return False
    return True
```

HideakiImamura commented:
Thanks for all reviews. Through addressing reviews, I've started to think the `_dominates_or_equal()` method does not need. We can simply replace all `_dominates_or_equal(p1, p2)` calls with `(p1 <= p2).all()`. It is more intuitive to replace it. Thanks, @toshihikoyanase!

HideakiImamura commented:
@hvy @toshihikoyanase PTAL

hvy commented on optuna/multi_objective/_hypervolume/wfg.py:
```diff
@@ -0,0 +1,95 @@
+            returned_limited_solution_set = [limited_solution_set[0]]
+            left = 0
+            right = 1
+            while left < right < n_points_of_s:
```
A detail but `left` is never larger than `right`.
```suggestion
            while right < n_points_of_s:
```

hvy commented on optuna/multi_objective/_hypervolume/wfg.py:
```diff
@@ -0,0 +1,95 @@
+        self._reference_point = None
+
+    def compute(self, solution_set: np.ndarray, reference_point: np.ndarray) -> float:
+        self._reference_point = reference_point
```
We you going to validate the arguments that the reference point is dominated by all points?

HideakiImamura commented on optuna/multi_objective/_hypervolume/wfg.py:
```diff
@@ -0,0 +1,95 @@
+        """
+        n_points_of_s = solution_set.shape[0]
+        dim = point.shape[0]
+        limited_solution_set = []
+
+        for i in range(n_points_of_s):
+            # If `point` is equal to or dominated by `solution_set[i]`,
+            # the returned solution set is `{point}`.
+            if (solution_set[i] <= point).all():
+                return point.reshape((1, dim))
+            limited_solution_set.append(np.maximum(solution_set[i], point))
+        limited_solution_set = np.asarray(limited_solution_set).reshape(
+            (len(limited_solution_set), dim)
+        )
```
I want to early return this function if `(solution_set[i] <= point).all()`, so it cannot be made one-line.

HideakiImamura replied:
It's my mistake. I adopt your suggestion. It will speed up the WFG algorithm. Thank you for your help!

HideakiImamura commented:
Thank you for the quick review @hvy! PTAL.

hvy commented on optuna/multi_objective/_hypervolume/base.py:
```diff
@@ -0,0 +1,35 @@
+        # Validates that all solution dominates or equal to the reference point.
+        for solution in solution_set:
+            if not (solution <= reference_point).all():
+                raise ValueError(
```
How about covering this case with unit tests?

hvy commented on tests/multi_objective_tests/hypervolume_tests/test_wfg.py:
```diff
@@ -0,0 +1,46 @@
+
+def test_invalid_input() -> None:
+    r = np.ones(3)
+    s = 2 * np.ones(3)
```
Should `s` not be 2d? Maybe it's overkill but we could also validate the input shapes.
```suggestion
    s = np.atleast_td(2 * np.ones(3))
```

HideakiImamura replied:
Ops! Thank you for the pointing. PTAL.

hvy replied:
Sorry for my obvious typo... and thanks for your quick action!

Crissman commented on optuna/multi_objective/_hypervolume/base.py:
```diff
@@ -17,4 +16,20 @@ def compute(self, solution_set: np.ndarray, reference_point: np.ndarray) -> floa
+
+    @staticmethod
+    def _validate(solution_set: np.ndarray, reference_point: np.ndarray) -> None:
+        # Validates that all solution dominates or equal to the reference point.
```
```suggestion
        # Validates that all solutions dominate or are equal to the reference point.
```

hvy replied:
This sentence seems to be [addressed already](https://github.com/optuna/optuna/pull/1537/files#diff-e400f5a3ed3f514c0e87fbbd588f1774R24). The following `ValueError` message could be fixed though, similar to what you suggest, i.e. "All solutions must dominate or be equal to ...".

HideakiImamura replied:
Yes, it is fixed as 
```python
# Validates that all points in the solution set dominate or equal the reference point.
```
Thank you for the comments.

HideakiImamura commented:
Thanks, @toshihikoyanase! PTAL.

toshihikoyanase commented on optuna/multi_objective/_hypervolume/wfg.py:
```diff
@@ -0,0 +1,89 @@
+        else:
+            # Assume limited_solution_set is sorted by its 0th dimension.
+            # Therefore, we can simply scan the limited solution set from left to right.
+            # Let `S'' = limited solution set`.
+            # `S''[left] is dominated by S''[right]` is never going to happen.
```
Could you remove this comment since it is not consistent with the comment L69 to L73 (almost Pareto optimal).
```suggestion
```

toshihikoyanase reviewed (changes requested):
Thank you for your update. It basically looks good to me.
Could you fix the code comment about the logic of `_limit`?

toshihikoyanase reviewed (approved):
Thank you for completing the long-running PR. LGTM.

I think we have possible follow-up tasks, so let me share them:
- Update documentation for developers. For example, we need to state `WFG` supports minimization only.
- More test cases. For example, the current test cases do not consider duplicated solution points, but they can be seen in the application (e.g., when evaluating the same parameter set with different workers).
